---
title: "RAD-seq: GBS on mice"
author: "Lucas A. Nell"
date: "March 31, 2016"
output:
  html_document:
    highlight: haddock
    theme: spacelab
  pdf_document:
    highlight: haddock
    latex_engine: xelatex
geometry: margin=1in
fontsize: 12pt
mainfont: Helvetica
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

# Demultiplex

I used `process_radtags` from [Stacks](http://catchenlab.life.illinois.edu/stacks/).

```{bash demx}
# Path to fastq files that need demultiplexed
cd /path/to/fastq

# Run number
export runNum=132571

# Restriction enzyme (we're using GBS so only have one)
export rEnz=apeKI

# Path to Stacks' `process_radtags` file (this will change depending on system)
export process_radtags=/usr/local/apps/stacks/latest/bin/process_radtags
# Tab-delimited text file with 3' barcode, 5' barcode reverse complement, and sample name
export barcodes=musGBS_${runNum}_bc.txt

# Files to demultiplex
export readOne=musGBS_${runNum}_1.fastq.gz
export readTwo=musGBS_${runNum}_2.fastq.gz

# Log output
export logFile=demultiplex_${runNum}.log

# Make "indiv" folder if it doesn't already exist
if [ ! -d "indiv" ]; then
    mkdir indiv
fi

# Now run
${process_radtags} \
    -1 ${readOne} \
    -2 ${readTwo} \
    -b ${barcodes} \
    -o ./indiv/ \
    --inline_inline --disable_rad_check \
    -e ${rEnz} \
    -r -i gzfastq &> ${logFile}
```

# Housekeeping

> *Note*: Renaming was only done to conform to other fastq files in my directory. This is
> unnecessary if you don't mind the *.1.fq *.2.fq naming from Stacks.

```{bash housekeeping}
cd /path/to/fastq/indiv

# Uncompress
for f in *.fq.gz; do gunzip $f; done

# Rename
for f in *.fq; do mv $f ${f/.fq/.fastq}; done
for f in *.1.fastq; do mv $f ${f/.1/_1}; done
for f in *.2.fastq; do mv $f ${f/.2/_2}; done
```


Because `process_radtags` adds "_2" to the end of read names in the R2 files, I'm going
to change them to match the read names in the R1 files. The old R2 files with the 
original `process_radtags` names are output into a folder named `"old"`.

```{bash houskeeping_2}
if [ ! -d "old" ]; then
    mkdir old
fi

for f in `ls *_2.fastq`
do
    ff=${f%2.fastq}fixed_2.fastq
    sed 's/_2$/_1/g' ${f} > ${ff}
    mv ${f} ./old/
    mv ${ff} ${f}
done

# Re-compress them (this will take a while!)
for f in *fastq; do gzip $f; done
```

# Aligning

Aligning was done using `bowtie`, the same as for DNA-seq data. See `basic_DNA-seq.Rmd`
for details.



# Stacking

## Create stacks

First we had to create stacks from the aligned BAM files using Stacks' `pstacks`. Because
BAM names included a unique integer after the first underscore, I used that for the id.

```{bash sstack}
cd /path/to/bam
export numThreads=12
export outPath=/path/to/stacks/

for bam in *.bam
do
    tmp=($(echo ${bam/%.bam/} | tr '_' ' '))
    id=${tmp[1]}
    /usr/local/apps/stacks/latest/bin/pstacks -t bam \
        -f ${bam} \
        -o ${outPath} \
        -i ${id} \
        -p ${numThreads} &> ./logs/pstacks_${id}.log
done
```

## Create stack catalog

Creating catalog using `cstacks`.

```{bash cstack}
cd /path/to/stacks
export numThreads=12
export outPath=./catalog/

pStacks=(`ls *.tags.tsv.gz | sed 's/.tags.tsv.gz//g'`)
export pStacks=${pStacks[@]/#/' -s '}

/usr/local/apps/stacks/latest/bin/cstacks \
    -b 1 \
    ${pStacks} \
    -o ${outPath} \
    -p ${numThreads} &> ${outPath}/catalog.log
```



## Search all samples against catalog

Search all samples from the `pstacks` step against catalog, using `sstacks`.

```{bash sstack}
cd /path/to/stacks
export numThreads=12
export outPath=./sets/
export catalogFile=./catalog/batch_1

pStacks=(`ls *.tags.tsv.gz | sed 's/.tags.tsv.gz//g'`)
export pStacks=${pStacks[@]/#/' -s '}

/usr/local/apps/stacks/latest/bin/sstacks \
    -b 1 \
    -c ${catalogFile} \
    ${pStacks} \
    -o ${outPath} \
    -p ${numThreads} &> ${outPath}/search.log
```

